commit 7f9346e11e160d0157786c1e3761453c7e0f99c4
Author: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date:   Sun Sep 6 21:07:00 2015 +0200

    Fix rules generating headers in hurd/ and mach/
    
    when initial make call has subdir= explicitly set.
    
    * sysdeps/mach/Makefile ($(patsubst
    mach%,m\%h%,$(mach-before-compile))): Force subdir to mach when
    calling $(MAKE).
    * sysdeps/mach/hurd/Makefile ($(patsubst %,$(hurd-objpfx)hurd/%.%,auth
    io fs process)): Force subdir to hurd when calling $(MAKE).
    ($(common-objpfx)hurd/../mach/RPC_task_get_sampled_pcs.c): Force
    subdir to mach when calling $(MAKE).

commit 697ed91ca901f8e2ce3ba2b7cf69cdb18c8b37ad
Author: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date:   Tue Sep 8 01:26:01 2015 +0200

    Fix parallel build of before-compile targets.
    
    * sysdeps/mach/Makefile ($(patsubst
    mach%,m\%h%,$(mach-before-compile))): Move rule to dedicated
    mach-before-compile target.
    * sysdeps/mach/hurd/Makefile ($(patsubst %,$(hurd-objpfx)hurd/%.%,auth
    io fs process)): Move rule to dedicated hurd-before-compile target.

--- a/sysdeps/mach/Makefile
+++ b/sysdeps/mach/Makefile
@@ -42,8 +42,10 @@ mach-before-compile := $(mach-objpfx)mac
 ifneq (mach,$(subdir))
 # This patsubst generates patterns like `m%h-shortcuts.h', which are damn
 # likely to match just the corresponding particular file we want.
-$(patsubst mach%,m\%h%,$(mach-before-compile)): # Run only if doesn't exist.
-	$(MAKE) -C $(..)mach mach-before-compile no_deps=t generating=t
+$(patsubst mach%,m\%h%,$(mach-before-compile)): mach-before-compile # Run only if doesn't exist.
+.PHONY: mach-before-compile
+mach-before-compile:
+	$(MAKE) -C $(..)mach subdir=mach mach-before-compile no_deps=t generating=t
 
 before-compile += $(mach-before-compile)
 endif
--- a/sysdeps/mach/hurd/Makefile
+++ b/sysdeps/mach/hurd/Makefile
@@ -46,15 +46,16 @@ hurd-objpfx = $(common-objpfx)hurd/
 
 # These are all the generated headers that <hurd.h> includes.
 before-compile += $(patsubst %,$(hurd-objpfx)hurd/%.h,auth io fs process)
-$(patsubst %,$(hurd-objpfx)hurd/%.%,auth io fs process): \
-  $(common-objpfx)mach/mach-shortcuts.h
-	$(MAKE) -C $(..)hurd before-compile no_deps=t
+$(patsubst %,$(hurd-objpfx)hurd/%.%,auth io fs process): hurd-before-compile
+.PHONY: hurd-before-compile
+hurd-before-compile: $(common-objpfx)mach/mach-shortcuts.h
+	$(MAKE) -C $(..)hurd subdir=hurd before-compile no_deps=t
 endif
 
 # Hurd profil.c includes this file, so give a rule to make it.
 ifeq ($(subdir),gmon)
 $(common-objpfx)hurd/../mach/RPC_task_get_sampled_pcs.c:
-	$(MAKE) -C $(..)mach before-compile no_deps=t
+	$(MAKE) -C $(..)mach subdir=mach before-compile no_deps=t
 endif
 
 
